File: stack.go
package main

import (
	"errors"
	"fmt"
)

type StackNode struct {
    value string
    next *StackNode
}

type Stack struct {
    head *StackNode
}

func createStack() *Stack {
    return &Stack{
        head: nil,
    }
}

func destroyStack(stack *Stack) {
    for stack.head != nil {
        //temp := stack.head
        stack.head = stack.head.next
        //temp = nil
    }
    stack = nil
}

func push(stack *Stack, value string) {
    newNode := &StackNode{
        value: value,
        next: stack.head,
    }
    stack.head = newNode
}

func pop(stack *Stack) (string, error) {
    if stack.head == nil {
        return "", errors.New("error")
    }
    
    futureDelete := stack.head
    value := futureDelete.value
    stack.head = stack.head.next
    futureDelete = nil
    return value, nil
}

func getStackElement(stack *Stack, index int) (string, error) {
    if index < 0 {
        return "", errors.New("error")
    }
    
    current := stack.head
    for i := 0; i < index; i++ {
        if current == nil {
            return "", errors.New("error")
        }
        current = current.next
    }
    
    if current == nil {
        return "", errors.New("error")
    }
    return current.value, nil
}

func printStack(st *Stack) {
    if st.head == nil {
        fmt.Println("Стек пуст.")
        return
    }

    fmt.Print("Стек: ")
    curElem := st.head
    for curElem != nil {
        fmt.Print(curElem.value + " ")
        curElem = curElem.next
    }

    fmt.Println()
}

File: stack.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <vector>
#include "stack.h"

using namespace std;

void createStack(Stack*& st) {
    st->head = nullptr;
}

SNode* createSNode(string data) { 
    SNode* newNode = new SNode{data, nullptr};
    return newNode;
}

void push(Stack* st, string data) {
    SNode* newNode = createSNode(data);
    newNode->next = st->head;
    st->head = newNode;
}

void pop(Stack* st) {
    if (st->head == nullptr) {
        cout << "Стек пуст.\n";
        return;
    }

    SNode* deleteNode = st->head;
    st->head = st->head->next;
    delete deleteNode;
}

void printStack(Stack* st) {
    if (st->head == nullptr) {
        cout << "Стек пуст.\n";
        return;
    }

    cout << "Стек: ";
    SNode* curElem = st->head;
    while (curElem != nullptr) {
        cout << curElem->value << " ";
        curElem = curElem->next;
    }

    cout << endl;
}